src/ForwardedHttpMiddleware.cs

```diff
@@ -1,27 +1,885 @@
         }
 
         public Task Invoke(HttpContext context)
         {
+            ApplyForwarders(context);
             return _next(context);
         }
+
+        private void ApplyForwarders(HttpContext context)
+        {
+            StoreOriginalValues(context);
+
+            var checkFor = _options.ForwardedHttp.HasFlag(ForwardedHttp.For);
+            var checkBy = _options.ForwardedHttp.HasFlag(ForwardedHttp.By);
+            var checkProto = _options.ForwardedHttp.HasFlag(ForwardedHttp.Proto);
+            var checkHost = _options.ForwardedHttp.HasFlag(ForwardedHttp.Host);
+
+            var request = context.Request;
+            var forwarded = request.Headers[ForwardedHeader];
+            var success = ParseProxyValues.TryParse(forwarded, out var forwardedHeader);
+            if (!success)
+            {
+                _logger.LogWarning("Invalid forwarded header");
+                return;
+            }
+
+            var entryCount = forwardedHeader.ForwardedValues.Count;
+
+            // Apply ForwardLimit, if any
+            if (_options.ForwardLimit.HasValue && entryCount > _options.ForwardLimit)
+            {
+                entryCount = _options.ForwardLimit.Value;
+            }
+
+            var sets = forwardedHeader.ForwardedValues
+                .Select(x => new SetOfForwarders
+                {
+                    RemoteIpAndPortText = x.For,
+                    LocalIpAndPortText = x.By,
+                    Scheme = x.Proto,
+                    Host = x.Host
+                })
+                .Reverse()
+                .Take(entryCount)
+                .ToArray();
+
+
+            // Gather initial values
+            var connection = context.Connection;
+            var currentValues = new SetOfForwarders()
+            {
+                RemoteIp = connection.RemoteIpAddress,
+                RemotePort = connection.RemotePort,
+                LocalIp = connection.LocalIpAddress,
+                LocalPort = connection.LocalPort,
+                // Host and Scheme initial values are never inspected, no need to set them here.
+            };
+
+            var checkKnownIps = _options.KnownNetworks.Count > 0 || _options.KnownProxies.Count > 0;
+            var applyChanges = false;
+
+            var entriesConsumed = 0;
+            for (; entriesConsumed < sets.Length; entriesConsumed++)
+            {
+                var set = sets[entriesConsumed];
+                if (checkFor)
+                {
+                    // For the first instance, allow remoteIp to be null for servers that don't support it natively.
+                    if (currentValues.RemoteIp != null && checkKnownIps && !CheckKnownAddress(currentValues.RemoteIp))
+                    {
+                        // Stop at the first unknown remote IP, but still apply changes processed so far.
+                        _logger.LogDebug(1, "Unknown proxy: {RemoteIpAndPort}", currentValues.RemoteIp);
```

Comment:
Requests commonly need to pass through a fixed number/set of proxies. This is a situation that could also have an option. E.g. if you don't see exactly the expected number of proxies/forwarders, reject the request with a 400. 

Consider this anywhere you find unparsible content, or not enough entries.